### ISSUE TITLE
__clang_analyzer__ macro is never defined

### ISSUE THREAD

**Issue by dkrupp** [Ericsson/codechecker#1422]
There are 2 issues in the false positive guide in section
https://github.com/Ericsson/codechecker/blob/master/docs/false_positives.md#-alternative-implementations
The example goes like this

```cpp
#ifndef __clang_analyzer__
unsigned f(unsigned x) {
  return (x >> 1) & 1;
}
#else
unsigned f(unsigned x) {
  return (x / 2) % 2;
}
#endif
```

1) `__clang_analyzer__` macro is never seems to be defined even when we run clang sa!? So always the first is taken.
2)  `(x >> 1) & 1` can be modeled properly by clang (at least by clang 5.0) so the example is not so good.

What about putting this example instead?:

```cpp
//EX: Alternative implementations !!__clang_analyzer__ macro is not defined!!
#ifndef __clang_analyzer__
//original implementation
char* fantastic_strcpy(char *to,const char *from){
  char *to_start=to;
  while (*from != 0){
    *to=*from;
    to++;
    from++;
  }
  *to=*from;    
  return to_start;
}
#else
//Simplified implementation for the analyzer
char* fantastic_strcpy(char *to,const char *from){  
  return strcpy(to,from);  
}
#endif
void caller_fantastic_strcpy(){
  char txt[]="Hello World!";
  char *buf=(char*) malloc(5);    
  fantastic_strcpy(buf,txt);//err:Buffer overflow DETECTED
}
```

Another fundamental question is that how the programmer knows whether his function is "properly" modeled by the analyzer?

**Comment by whisperity** [comment]
**Please format your issue texts accordingly so that special characters like `_` are understood as code, and not as a Markdown formatting predicate.**

The guide most likely comes from the [official Clang False Positive guide](https://clang-analyzer.llvm.org/faq.html#exclude_code), which means that that *too* is outdated. The only findings for the word `__clang_analyzer__` in the Clang source code are:

```
15:33 whisperity@linux:~/llvm-release_60$ fse "__clang_analyzer__"
./tools/clang/lib/Frontend/InitPreprocessor.cpp:971:    Builder.defineMacro("__clang_analyzer__");
./tools/clang/test/Analysis/misc-ps.m:5:#ifndef __clang_analyzer__
./tools/clang/test/Analysis/misc-ps.m:6:#error __clang_analyzer__ not defined
./tools/clang/www/analyzer/faq.html:203:help us improve user experience. As the last resort, consider using <tt>__clang_analyzer__</tt> macro
./tools/clang/www/analyzer/faq.html:209:defines the preprocessor macro <tt>__clang_analyzer__</tt>. One can use this 
./tools/clang/www/analyzer/faq.html:213:#ifndef __clang_analyzer__
```

Apart from the documentation, the relevant snippets taken from current `master` (but these lines are the same as in LLVM 6.0)

https://github.com/llvm-mirror/clang/blob/3a6ae089a92143aa4cef318f8fccafc05f8b0e3d/lib/Frontend/InitPreprocessor.cpp#L969-L971

https://github.com/llvm-mirror/clang/blob/3a6ae089a92143aa4cef318f8fccafc05f8b0e3d/test/Analysis/misc-ps.m#L5-L7

**Comment by Xazax-hun** [comment]
In case the macro is not set the most likely cause is that we do not properly invoke the analyzer.

**Comment by irishrover** [comment]
This  macro was undefined for clang-tidy, while clang-sa defines it.
But in https://reviews.llvm.org/rCTE331474 I add this macro definition into clang-tidy too.

**Comment by whisperity** [comment]
The answer to the fundamental question is that there is no way to know. Perhaps generating and investigating an exploded graph, but that's my best hunch.

**Comment by whisperity** [comment]
@irishrover Thank you for that patch.

**Comment by gyorb** [comment]
For the ctu collect phase it was set too see: #1865

**Comment by johan-boule** [comment]
@irishrover Even on the latest release, clang-tidy 15.0.6, the __clang_analyzer__ macro is not predefined.
